core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java

```diff
@@ -0,0 +1,231 @@
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+
+  @Test
+  public void pickAfterResolvedAndUnchanged() throws Exception {
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+
+    verify(mockHelper, times(1)).createSubchannel(any(EquivalentAddressGroup.class),
+        any(Attributes.class));
+    verify(mockHelper, times(1)).updatePicker(any(Picker.class));
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterResolvedAndChanged() throws Exception {
+    SocketAddress socketAddr = new FakeSocketAddress("server3");
+    List<SocketAddress> newSocketAddresses = Lists.newArrayList(socketAddr);
+    List<ResolvedServerInfoGroup> newServers = Lists.newArrayList(
+        ResolvedServerInfoGroup.builder().add(new ResolvedServerInfo(socketAddr)).build());
+
+    final Subchannel oldSubchannel = mock(Subchannel.class);
+    final EquivalentAddressGroup oldEag = new EquivalentAddressGroup(socketAddresses);
+    when(oldSubchannel.getAddresses()).thenReturn(oldEag);
+
+    final Subchannel newSubchannel = mock(Subchannel.class);
+    final EquivalentAddressGroup newEag = new EquivalentAddressGroup(newSocketAddresses);
+    when(newSubchannel.getAddresses()).thenReturn(newEag);
+
+    when(mockHelper.createSubchannel(any(EquivalentAddressGroup.class), any(Attributes.class)))
+        .then(new Answer<Subchannel>() {
+          @Override
+          public Subchannel answer(InvocationOnMock invocation) throws Throwable {
+            Object[] args = invocation.getArguments();
+            EquivalentAddressGroup eag = (EquivalentAddressGroup) args[0];
+            if (eag.equals(oldEag)) {
+              return oldSubchannel;
+            } else if (eag.equals(newEag)) {
+              return newSubchannel;
+            }
+            throw new IllegalArgumentException();
+          }
+        });
+
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+    loadBalancer.handleResolvedAddresses(newServers, affinity);
+
+    verify(mockHelper, times(2)).createSubchannel(eagCaptor.capture(), eq(affinity));
+    verify(mockHelper, times(2)).updatePicker(pickerCaptor.capture());
+
+    assertEquals(socketAddresses, eagCaptor.getAllValues().get(0).getAddresses());
+    assertEquals(newSocketAddresses, eagCaptor.getAllValues().get(1).getAddresses());
+
+    Subchannel subchannel = pickerCaptor.getAllValues().get(0).pickSubchannel(
+        affinity, new Metadata()).getSubchannel();
+    assertEquals(oldSubchannel, subchannel);
+
+    Subchannel subchannel2 = pickerCaptor.getAllValues().get(1).pickSubchannel(affinity,
+        new Metadata()).getSubchannel();
+    assertEquals(newSubchannel, subchannel2);
+    verify(subchannel2, never()).shutdown();
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChangeBeforeResolution() throws Exception {
```

Comment:
Yes, resolution didn't happen so I'm just checking if there were any interactions with `mockHelper` - I'll rename it.